### ISSUE TITLE
[OpenGL] -g0 option strips information needed for vertex <> geometry shader interface matching.

### ISSUE THREAD

**Issue by hanfling** [KhronosGroup/glslang#2939]
As far as I understand it, the geometry shader requires you to use an input block, which in turn will require the elements inside the blocks to match by name. However the name is stripped with -g0 hence breaking interface matching.

Reference: https://www.khronos.org/opengl/wiki/Shader_Compilation#Interface_matching

**Comment by greg-lunarg** [comment]
Are you saying that if you compile without -g0 that your application works and if you compile with -g0 that your application does not work?

Can you please provide sample shaders and the symptoms you are seeing that suggest the SPIR-V is in error? Also provide any details regarding how you are compiling the shaders.

**Comment by hanfling** [comment]
> Are you saying that if you compile without -g0 that your application works and if you compile with -g0 that your application does not work?

Yes, more specifically I'm using vert/geom/frag SPIR-V shaders. As soon as I use -g0 in the vert stage linking the program fails. One exception is one vert/geom/frag combination I'm using which works with -g0 on all shader stages, but I suspect that the fallback names just line up there.

> Can you please provide sample shaders and the symptoms you are seeing that suggest the SPIR-V is in error? Also provide any details regarding how you are compiling the shaders.

I was afaid someone would ask for that 🍡 

---

Build commands I use. For building without -g0 option it's the same command with just that option removed.

	Bin\glslangValidator.exe -g0 -G100 -S vert --vn EndFlashVertexShaderBinary -o Shaders\Compiled\EndFlashVertexShader.h Shaders\EndFlashVertexShader.glsl
	Bin\glslangValidator.exe -g0 -G100 -S geom --vn EndFlashGeometryShaderBinary -o Shaders\Compiled\EndFlashGeometryShader.h Shaders\EndFlashGeometryShader.glsl
	Bin\glslangValidator.exe -g0 -G100 -g0 -S frag --vn EndFlashFragmentShaderBinary -o Shaders\Compiled\EndFlashFragmentShader.h Shaders\EndFlashFragmentShader.glsl

---

OpenGL Implementation used for testing:

	GL_VENDOR:   NVIDIA Corporation
	GL_RENDERER: GeForce GTX 960/PCIe/SSE2
	GL_VERSION:  4.6.0 NVIDIA 461.40

I'm just linking the three shader stages into a non separable program. The error messages below are all info log messages after glLinkProgram failed (GL_LINK_STATUS==GL_FALSE).

---

As mentioned above -g0 in both frag and geom shader works fine.

Using -g0 on vert only results in following error:

	Link info
	---------
	error: Block "_Input" not declared as an output from the previous stage

---

One (simple) combination of shaders I'm using which will fail:

[EndFlashGeometryShader.glsl.txt](https://github.com/KhronosGroup/glslang/files/8633502/EndFlashGeometryShader.glsl.txt)
[EndFlashVertexShader.glsl.txt](https://github.com/KhronosGroup/glslang/files/8633503/EndFlashVertexShader.glsl.txt)
[EndFlashFragmentShader.glsl.txt](https://github.com/KhronosGroup/glslang/files/8633504/EndFlashFragmentShader.glsl.txt)

**Comment by hanfling** [comment]
Update: Missed an include file above [and github complains about file extension when I try to edit post above].

[OpenGLShaderStructures.h.txt](https://github.com/KhronosGroup/glslang/files/8636782/OpenGLShaderStructures.h.txt)

**Comment by hanfling** [comment]
Update: Missed include files above [and github complains about file extension when I try to edit post above].

[OpenGLShaderBindings.h.txt](https://github.com/KhronosGroup/glslang/files/8636795/OpenGLShaderBindings.h.txt)
[OpenGLShaderStructures.h.txt](https://github.com/KhronosGroup/glslang/files/8636796/OpenGLShaderStructures.h.txt)

**Comment by greg-lunarg** [comment]
Sorry for the delay.

I am thinking that your vertex and geometry shader interfaces do not match according to OpenGL rules. From the spec:

> Outputs and inputs can either be in an [Interface Block](https://www.khronos.org/opengl/wiki/Interface_Block) or as loose output/input variables. The rules for matching differ between them.

Then follows rules for how blocks match blocks  and how loose variables match loose variables. There are no rules for how loose variables match a block.

In general, matching by names is not supported when using SPIR-V with OpenGL. Interface matching must be done by following the loose variable rules ie location and type and qualifier matching.

The driver is being permissive in using debug info to help match your output and input here. Stripping the debug information only reveals the problem with your shaders that I mentioned above.

So for your particular case, I think that creating a block in the vertex shader for your outputs would fix your problem.

If you believe my analysis is in error, please respond and I will re-open.

**Comment by hanfling** [comment]
I looked at the issue again and actually figured out by trying how to not use an input block for the geometry shader, so I can match over the rules which won't require a name. So that was certainly a missconception on my end.

And yeah I saw the section on SPIR-V interface matching in the section now. Your point is valid that the implemention should not use the names.

Alright, thanks for your time :)
